### ISSUE TITLE
Any roadmap or planning for PHP 8 support ?

### ISSUE THREAD

**Issue by pounard** [eclipse/pdt#95]
Hello,

Is there any plans/planned date for supporting PHP 8 ?

Best regards, and thank you for all the hard work.

**Comment by pounard** [comment]
I should add that I'm very interested in having it, since I'm going to use it from approximately now for projects.

**Comment by zulus** [comment]
It was planned during holiday, but second covid wave happen... I'll grab some time during Christmas and prepare a least basic PHP8 support (syntax checker). 

I'll also need this version soon ;)

**Comment by pounard** [comment]
@zulus I have a question about contributing, are you modifying the CUP parser files manually for upgrading the parser ?

And also, are you are parsing types using the `Identifier` class, but with union type, you'll need to a case in the `type::` grammar, case in which you'll need to parse an `IdentifierList` or some sort, I started reading the code, they are thousands of references of the `Identifier` class, seems hard to grasp where you use it for types and where you use it for other expressions, are you planning on specializing a bit more parsed tokens in the AST ?

I stopped at this point on sunday night, I read thousands of lines of code in PDT, I'm interesting in helping is that's possible, I'll probably ask many questions thought.

**Comment by zulus** [comment]
If you modify CUP file, eclipse regenerate parser during build. You can even find eclipse plugin with basic colouring/validation.

As far as I remember PHP8 require from us a lot of AST and model changes, or multiple string reparsing during code analysis (performance degradation). Due this, in last year I started working on completely fresh model based on eclipse handly, without DLTK garbage. But due world madness I haven't time to finish this.

For example DLTK model require return type as simple string. So, to support union during inference, validation, code assist etc..., split string has to be done in many different places.

**Comment by pounard** [comment]
> For example DLTK model require return type as simple string. So, to support union during inference, validation, code assist etc..., split string has to be done in many different places.

Ouch.

> As far as I remember PHP8 require from us a lot of AST and model changes

That's what I thought when I saw your current parser, I guess union types is the most invasive change, you'll need to adapt your AST too, if not for performance, at least it must be done for maintainability.

**Comment by zulus** [comment]
We also have two ASTs:

1. Advance, as legacy form JDT, with basic binding, cache etc. Used by semantic colouring (like deprecated) and by some content assist features (like ctrl/cmd+2 L)
2. Compiler, based od DLTK required for inferencing, validation, navigation etc...

Most of things we can adopt here, main problem is DLTK. Lot of thing have to be override and/or re-parsed later. This madness has to be stopped, this is why I started playing with handly.

**Comment by pounard** [comment]
In an ideal world, PHP would ship with a language server using its own parser, but we can only dream of it :) Just out of curiosity, did you consider switching over a LSP based editor ? There's a few PHP LSP implementations existing, some editors use those, but I have no idea how advanced they are.

**Comment by zulus** [comment]
In general, what I want/plan:

1. Drop DLTK - copy/migrate lucene search and code inference 
2. Drop WTP - use WildWebDeveloper only
3. Use GenericEditor (LSP ready)
4. Merge Doctrine/Twig/Symfony plugins into core as example howto extend PDT
5. In future deliver LSP server based on PDT, similar as JDT doing

PDT have one leg in WTP (old times), second in DLTK. ~20-30% of coding now is workaround between two worlds

I didn't plan to use existing LSP, because protocol in general wasn't designed to create extensions, (for example introduce code assist for `$em->getRepository()`

**Comment by pounard** [comment]
OK, it seems a very nice plan, not so sure I can in those conditions I don't know the Eclipse platform and tooling enough for anything here... But, once the huge refactor is done, I'll probably be able to fix some bugs here and there :)

**Comment by zulus** [comment]
> OK, it seems a very nice plan, not so sure I can help in anything in those conditions I don't know the Eclipse platform and tooling enough for anything here... But, once the huge refactor is done, I'll probably be able to fix some bugs here and there :)

In many cases it's rewrite ;)

**Comment by pounard** [comment]
Hey that said, I'm OK for being a beta tester of yours, if you wish, I'm very eager to test and help, and report bugs, that's what I do best :fuelpump:

**Comment by pounard** [comment]
Is this still alive ?

**Comment by rickyrocker** [comment]
yip - I'm with @pounard - is there any activity here or has the project stalled.

If so, is there any way I can help move it forward as it is such a great tool it would be a tragedy to let it die - I'm a PHP developer so not sure if useful but happy to help any way I can

**Comment by stanolacko** [comment]
I'm also in.

**Comment by rickyrocker** [comment]
Have just received 

  PHP Development Tools (PDT)	8.0.0.202007171531	org.eclipse.php.feature.group	Eclipse PDT
  
How epic is that!!  thank you!

(now if I can just manage to work out how to get my project language library from PHP 7.4 to PHP 8.0)

**Comment by pounard** [comment]
@rickyrocker I'm using 8.x also, but there is no PHP 8 support.

**Comment by rickyrocker** [comment]
eish - thanks @pounard

@zulus is there any way I can help move this forward? having named parameters is literally game changing for me.

**Comment by rickyrocker** [comment]
@zulus - i really don't want to hassle but it would be incredibly good if you could provide at least some form of status here. This project is I'm sure of value to a great many PHP developers and it really doesn't seem right just to leave everyone hanging regarding the way forward.

Really looking forward to hearing from you!

thanks in advance

**Comment by blackjyn** [comment]
PHP 8 support please.... Its a shame why there are no clear info on this.
at least , show the roadmap and give some words y committee!

**Comment by zulus** [comment]
OK, I finally grab some time...

**Comment by rickyrocker** [comment]
@zulus excellent- I'm very very happy to help if I can, but would appreciate some guidance as to how best to go about getting involved

**Comment by tbl0605** [comment]
> 
> 
> OK, I finally grab some time...

@zulus, I'll also try to grab some time the second semester, we could start by adding support for PHP 8.0 in our parsers&lexers ;)
Since I'm not using PHP 8.0 (yet), it would be nice to list all new language features that should go in our parsers&lexers (and code formater), so at least users could use PHP 8.0 syntax. 
What do you think?
Maybe in some separate issue(s)?

**Comment by rickyrocker** [comment]
@zulus  @tbl0605  any updates - named parameter code completion remains game changing for us!

**Comment by pounard** [comment]
Actually all features of PHP 8 are game changing, I can't wait having a proper support of union types, match expression and attributes.

**Comment by tbl0605** [comment]
> 
> 
> @zulus @tbl0605 any updates - named parameter code completion remains game changing for us!

On my side, I can have a deeper look only end of June. But without any answer of @zulus I won't be able to be a big help ;)
Anyway we're only 2 developers left on PDT for now, so we wll sadly be less responsive than one or two years ago.

**Comment by pounard** [comment]
@tbl0605 I understand that, being a contributor to open source is hard. But the question we all have on our list is do you have a clear roadmap if not a deadline for this ?

**Comment by rickyrocker** [comment]
@tbl0605 @zulus  we are also willing to look at helping fund development but I can't see any way to achieve this

**Comment by tbl0605** [comment]
> 
> 
> @tbl0605 I understand that, being a contributor to open source is hard. But the question we all have on our list is do you have a clear roadmap if not a deadline for this ?

Not yet for PHP 8.0, we already have done feature roadmaps for previous versions of PHP, so it should not be a big deal. Our "time deadline" was (until recently) based on Eclipse's release cycles (every 3 months at best or 6 months at worst). 
Maybe full support for PHP 8.0 will be made accross multiple versions of eclipse.
Again, it's only a supposition, I really don't know what Dawid is planning (or has already done) ;)

**Comment by tbl0605** [comment]
> 
> 
> @tbl0605 @zulus we are also willing to look at helping fund development but I can't see any way to achieve this

Thank you, that's great to know! :) Personaly I never worked on open source projects using helping funds, so I'm pretty new in this area. I know it can be done through github sponsoring. For sure we should do that for PDT, it would maybe attract new developers, or maybe help Dawid on its PDT work.

**Comment by zulus** [comment]
Unfortunately Handly based model is still in experiment phase :/

I'll introduce PHP 8 parser for current model in next week

**Comment by o-alquimista** [comment]
Did any of the PHP 8 features ship with the version 2021-06 that was released this month? My distro (Fedora) is still on 2021-03, so I need to decide whether to ditch this and get the latest.

Named parameters, maybe? This one would be really nice, because currently the IDE thinks it's an error in the code.

**Comment by pounard** [comment]
@o-alquimista I just upgraded, still no PHP 8 support.

**Comment by pounard** [comment]
Hello, believe me, I really don't want to be rude or anything and I do highly respect the work you do, but is PHP 8 support going to land soon, or one day, is there at least an approximate timeline for this ?

**Comment by rickyrocker** [comment]
Message to the project team.

I understand completely that development when no being paid is not a cake walk and I certainly give thanks for the effort so far, however the reality is this:

1. It would be simple to keep end-users realistically updated as to intention: Your communication completely sucks.
2. Point 1 has a very real cost to businesses depending on the IDE: it removes the ability to determine a strategy and reduces the ability to take advantage of new language features, resulting in longer times to create suboptimal code.

So, please do the right thing and make a clear statement regarding what the intention is to support PHP 8, and in what timeframe.

**Comment by rickyrocker** [comment]
From the complete lack of response after nearly a month, I can only assume this project has gone the way of the dodo and we will need to migrate to visual studio code.

**Comment by rimace** [comment]
@zulus maybe a crowd-founding campaign could help?
I mean I am willing to pay for PHP 8 support and I thing at least a few would spend a few bucks on that feature. Just an idea if someone steps up and would like to do it for some pocket money.

**Comment by rickyrocker** [comment]
A few bugs aside PDT is such a great PHP IDE and it is a tragedy that there is no apparent intention to support PHP 8. 

VS Code does, but as far as I'm concerned it isn't a patch on the Eclipse IDE and we are really loathe to move. We're not a huge company but we would happily contribute an amount each month to keep developers on this project if we had some understanding that there was commitment.

I simply can't believe there aren't more people upset about this. And I simply can't believe that funding is not possible to get for the developers. 

We just need to know how.

**Comment by zulus** [comment]
I’m sorry. I finished worst and longest crunch ever. I’ll refresh tasks soon :/

**Comment by rickyrocker** [comment]
> 
> 
> I’m sorry. I finished worst and longest crunch ever. I’ll refresh tasks soon :/

@zulus - can we not help fund in some way? I would need to confirm this internally but we have 5 developers currently using the project, and I'm fine to pay an annual amount for each developer that would roughly equate with what paid tools such as say PHPstorm charges (I would like 50% of our funding to go directly to this project) - this could be win:win as the problem for us with the free model is that the project is always "second" as long as developers are unpaid

**Comment by pounard** [comment]
@zulus you've been the missing person that made me the most anxious of the year :) Just kidding, I'm really pleased to see you back. If we can help in any way, please just ask !

**Comment by pounard** [comment]
I'm sorry it I'm getting annoying, but is it planned sometime soon ? Now we have PHP 8.1 released, I'm so excited being able to use those features without breaking my editor someday ^^

**Comment by zulus** [comment]
For now I fixed compatibility issue between PDT and upcoming Eclipse 2021-12. Together with 2021-12, PDT 7.2.1 will released (couple fixes).

After that I'll continue my work on 8.0/8.1. Have to be done for 2022-03

**Comment by blackjyn** [comment]
thank you @zulus for the confirmation! Best luck!

**Comment by rickyrocker** [comment]
Hope this is still on track @zulus  - I'm so looking forward to named parameters!

**Comment by zulus** [comment]
Yeah, work in progress ;)

**Comment by pounard** [comment]
@rickyrocker union types, mixed, match expression and null coalesce operator :D !!!

**Comment by pounard** [comment]
And now that 8.1 us hère, readonly properties !

**Comment by rickyrocker** [comment]
March 16, 2022 !!

**Comment by mlocati** [comment]
> March 16, 2022 !!

What's this date?

**Comment by rickyrocker** [comment]
> @rickyrocker union types, mixed, match expression, attributes and null coalesce operator :D !!!

yes, have just come across union types - will be great to see them!

**Comment by rickyrocker** [comment]
@zulus  : Is the release still on track? (if not, please do say so as our expectation levels are very high and the sooner we know if problems the better - thanks)

**Comment by pounard** [comment]
@zulus and please consider us being able to alpha/beta test anything, if you need.

**Comment by rickyrocker** [comment]
Here is the [PDT development project on Github](https://github.com/eclipse/pdt)

It looks pretty quiet to me but I'm not sure I'm understanding it correctly.

@zulus please please can you confirm whether the project is on track to release 8 support in 2022-03 

This is very important for us as we have started writing 8 code in anticipation of the upgrade so we are forced to ignore editor errors/warnings - which we won't be able to sustain for long.

**Comment by agmckee** [comment]
@rickyrocker I think Dawid (zulus) develops on his own machine and then pushes up to his clone when he has something ready for PR or merge. It's just Dawid working on PHP 8 support. That's why the repo is quiet.

**Comment by pounard** [comment]
@rickyrocker I guess not, I guess it's just not released yet.

**Comment by agmckee** [comment]
@pounard I checked earlier and saw the same as you but the downloads page is now showing 2022-03 for me so it looks like they have released it now.

@rickyrocker there's an Eclipse repo for PDT 8.0 at https://download.eclipse.org/tools/pdt/updates/8.0/

**Comment by mlocati** [comment]
Since this repository ( https://github.com/eclipse/pdt ) contains the source code of PDT, it's rather easy to check if it gets updated: simply go to https://github.com/eclipse/pdt/commits/master to know the latest changes

**Comment by rickyrocker** [comment]
> @pounard I checked earlier and saw the same as you but the downloads page is now showing 2022-03 for me so it looks like they have released it now.
> 
> @rickyrocker there's an Eclipse repo for PDT 8.0 at https://download.eclipse.org/tools/pdt/updates/8.0/ Unfortunately it appears to be the same as the last release, so not sure why it exists. There does not seem to be PHP 8.0 support of any kind. I suspect this project is really dead at this point, I am looking for an alternative myself.

I've got 8.0 installed thanks and likewise see nothing

It does rather look like this project is dead and if so - while I have no issue with @zulus  stopping development due to lack of funding or time - I am VERY unimpressed with him for stringing businesses along for so long. As I have explained in the past to him, there are very real financial implications associated with this. We have held off on migrating and upgrading to PHP 8 based solely on his commitments. 

@zulus  - very very poor.

I can tell you an alternative is Visual Studio Code which has supported PHP 8 for some time now.

**Comment by pounard** [comment]
Please don't insult the maintainer, it's not acceptable. People have a life, and they don't owe you anything.

**Comment by rickyrocker** [comment]
> @rickyrocker I guess not, I guess it's just not released yet. https://www.eclipse.org/downloads/ is still not advertising 2022-03 so I guess it's not even released stable yet.

@pounard  thanks - I've installed 2022-03 and still seeing only 7.4 max in the list of possible PHPs

**Comment by rickyrocker** [comment]
> Please don't insult the maintainer, it's not acceptable. People have a life, and they don't owe you anything.

I will absolutely insult him if he has promised delivery and just left us hanging. I don't expect work but I do expect honesty and candour - and I have a right to expect that from everyone at no charge.

**Comment by pounard** [comment]
No actually you don't, sometime promises can't be kept, and I don't remember him promising anything anyway. It's just plain disrespectful. You have the right to be disappointed, we all here are I guess, but we are patient and nice, because that's how open source is done, not the other way around. There's many open source communities where you'd be ban for this, legitimately. You should at least be happy that there's not much active people here that are able to do so.

**Comment by rickyrocker** [comment]
if I get banned for being honest and expecting the same in return - so be it.  ...I will not be the loser.

**Comment by agmckee** [comment]
> Please don't insult the maintainer, it's not acceptable. People have a life, and they don't owe you anything.

Dawid [publicly committed to delivering PHP 8.0 support for this release](https://github.com/eclipse/pdt/issues/95#issuecomment-983883013), which was unwise if he couldn't deliver. I'm not particularly impressed myself, I've offered my help several times. PDT is a product of many people's effort, if Dawid is unable to keep the codebase up to date with current PHP releases he should say so to the Eclipse Foundation so that other developers can be found. Or he could work with people who have offered their help so that they can step in and work on PHP 8.0 support. I have been looking into adding this support myself, which would be considerably easier if Dawid would work with me to help point me in the right direction.

Saying things like:

> After that I'll continue my work on 8.0/8.1. Have to be done for 2022-03

... and not delivering is a sure fire way to irritate users. Just because it's open source doesn't mean there's no expectation of communication.

PHP 7.4 is end of life in 8 months time and at this point it doesn't look realistic that this project will provide PHP 8.0 support by that time.

**Comment by rickyrocker** [comment]
> > Please don't insult the maintainer, it's not acceptable. People have a life, and they don't owe you anything.
> 
> Dawid [publicly committed to delivering PHP 8.0 support for this release](https://github.com/eclipse/pdt/issues/95#issuecomment-983883013), which was unwise if he couldn't deliver. I'm not particularly impressed myself, I've offered my help several times. PDT is a product of many people's effort, if Dawid is unable to keep the codebase up to date with current PHP releases he should say so to the Eclipse Foundation so that other developers can be found. Or he could work with people who have offered their help so that they can step in and work on PHP 8.0 support. I have been looking into adding this support myself, which would be considerably easier if Dawid would work with me to help point me in the right direction.
> 
> Saying things like:
> 
> > After that I'll continue my work on 8.0/8.1. Have to be done for 2022-03
> 
> ... and not delivering is a sure fire way to irritate users. Just because it's open source doesn't mean there's no expectation of communication.
> 
> PHP 7.4 is end of life in 8 months time and at this point it doesn't look realistic that this project will provide PHP 8.0 support by that time.

Thank you! I 100% agree.

**Comment by pounard** [comment]
You are not this person, and you don't know anything about his life, so please at least, be disappointed in a respectful manner.

**Comment by rickyrocker** [comment]
> You are not this person, and you don't know anything about his life, so please at least, be disappointed in a respectful manner

@pounard  you are welcome to your perspective on what is acceptable and how people should engage - what is clear is that it is not the only one.

**Comment by pounard** [comment]
There are many ways to skin a cat, but lacking of respect is *never* a way do things.

**Comment by rickyrocker** [comment]
> There are many ways to skin a cat, but lacking of respect is _never_ the way to go.

I 100% agree with you, and that is exactly what I'm expressing my frustration @zulus for.

**Comment by rickyrocker** [comment]
> > There are many ways to skin a cat, but lacking of respect is _never_ the way to go.
> 
> I 100% agree with you, and that is exactly what I'm expressing my frustration @zulus for.

There are many ways to show respect ...one very important one is to have respect for the people you make commitments to. He has not and I have expressed my dissatisfaction. There is only one reason I could accept his lack of follow through and communication and that is if he has been unable to access his computer since his last commit. In that case @zulus  I apologise for my frustration. But if that is not the case, I don't. Period.

**Comment by rickyrocker** [comment]
guys - it's been fun but I don't see much point in this list as it is clearly devoid of progress. So @pounard I'll satisfy your moderation instinct and ban myself and leave you to wait patiently and with respect. Good luck.

Bye.

**Comment by pounard** [comment]
> PDT is a product of many people's effort, if Dawid is unable to keep the codebase up to date with current PHP releases he should say so to the Eclipse Foundation so that other developers can be found. Or he could work with people who have offered their help so that they can step in and work on PHP 8.0 support. I have been looking into adding this support myself, which would be considerably easier if Dawid would work with me to help point me in the right direction.

Yes I agree. From what I understood, he was planning a refactor before, I'd very much indeed to see PDT code more simple so I could step up and do occasional PR sometime. But right now, I guess we have to wait for him to put things in order.

I'd love to help, but we simply can't, at least until he resurface. That's sad but sometime, that's what happens with free software. If by chance I was an experienced Eclipse dev, I'd fork, but I don't have the competences to do so (last time I wrote Java was 15 years ago, last time I wrote Eclipse plugins 16 years, things have evolved since).

**Comment by tianfan007** [comment]
does this project  waw dead? the lastest version has been released about  two years, PDT dosen't support php 8 yet

**Comment by tianfan007** [comment]
does this project was dead? the lastest version has been released about two years, PDT dosen't support php 8 yet

**Comment by wakabadou** [comment]
There seems to be some progress, but no commits have been made.
Europeans are busy now, so thank you for zulus contribution and wait patiently.

**Comment by jeferod83** [comment]
> There seems to be some progress, but no commits have been made. Europeans are busy now, so thank you for zulus contribution and wait patiently.
> 
> PDT is the best PHP IDE for me. The faster delivery is a great pleasure.

For me too. I'm using Eclipse since 2008, and it is more heavy than other IDE's? Yes. I swear that I tried to use another IDE, but... Eclipse is Eclipse, and I love it.

**Comment by vardumper** [comment]
I can't really contribute anything here, but I am also impatiently waiting for PHP8.* support, too.

**Comment by pounard** [comment]
Any news on this ?

**Comment by trenshaw** [comment]
I'm running PHP 8.1.2 on Ubuntu 22.04 (default package with distro). Support for PHP 8.1 validation and syntax checking is becoming urgent for PDT to remain relevant. Let me know if there is anything I can do to help.

**Comment by ricky-rockstar** [comment]
I have sent the following mail to the Eclipse Foundation in the hope that they will provide us with some form of response. Let's see.

Hi,

There are a ton of developers who have been waiting probably over a year now for the lead developer of the PDT (PHP) project to provide any sort of update and people are becoming desperate. See https://github.com/eclipse/pdt/issues/95 

The problem is less about whether PDT is going to continue, and more about the lack of any form of formal communication to the user base as to what is going on.

I am one of these people – I love Eclipse and PDT but this not knowing is killing our business as we are unable to take advantage of any of the new features of PHP because Eclipse doesn’t support them.

I am therefore writing to you in the hope that you will engage the project leads, get an answer, and make a formal statement to the ever more desperate user base.

Thank you in advance.

**Comment by waynebeaton** [comment]
Sorry for the long delay in responding. I only just became aware of the email (I have only lame excuses).

To manage expectations, the EMO has no special powers to force committers to make specific commitments regarding the allocation of resources. Ultimately, Eclipse PDT is a community project and relies on contributions from the community to survive and thrive. 

The core problem here, AFAICT, is that the project is chronically under resourced.

We have a couple of people in our community who are good at organising willing communities and revitalising open source projects. I'll reach out to my contacts to see if I can get one of them interested in helping with the community development and planning aspects of Eclipse PDT, which would allow @zulus more space to focus on the core implementation. For this to work, we're going to need at least a few folks to contribute some development resources to the effort.

I'll also bring the project to the attention of the [Eclipse IDE Working Group](https://eclipseide.org/working-group/) to see if there's anything that they can do to help.

**Comment by NainzguL** [comment]
> For this to work, we're going to need at least a few folks to contribute some development resources to the effort.

I don't have a lot of time, but if few hours of my time can help PDT to support PHP 8, it worth every minute.
Let me know what can I do, and how can I help.

**Comment by stanolacko** [comment]
I'm in!

**Comment by vardumper** [comment]
If a PHP developer is of any help, then I'm in, as well. But probably I will be limited to testing and giving feedback.

**Comment by waynebeaton** [comment]
> We have a couple of people in our community who are good at organising willing communities and revitalising open source projects. 

Unfortunately, I have been unable to find any of our "usual suspects" who have cycles enough to assist.

I still have hope that the IDE Working Group may have some ideas on how to help. I'll report back the minute I hear anything from them.

**Comment by sebromon** [comment]
Dear all,
I would like to contribute to this project. I have some free time.
From what I understand, Dawid started a very long and useful refactoring work.
Maybe a temporary solution would be to fork the project and develop a basic support for PHP 8.
@zulus , can you commit your actual work in order to see if we can help you in this migration.

**Comment by pounard** [comment]
I have to admit, PHP 8 is becoming critical now that PHP 7.4 is not supported anymore, and I still fight because I love Eclipse UI, but being unable to use reorganize imports as soon as someone (or me) uses PHP 8 feature is becoming a real burden (for PHP 8 files I have to use my IDE as a basic notepad... because everything is red due to parser not understanding the code anymore).

**Comment by ungeahnt** [comment]
see @nabice 's announcement here:

https://github.com/eclipse/pdt/issues/141#issuecomment-1340710057

**Comment by ungeahnt** [comment]
Just out of interest: There have been offers to collaborate in recent months. Has there been any cooperation with @zulus in the meantime?

**Comment by zulus** [comment]
For now I'm alone. I'm trying to split things and publish :/

**Comment by merks** [comment]
@zulus  If there is anything I can personally do to help, e.g., with an Oomph setup for the project or with ci build or Tycho assistance, please feel free to ask help.

**Comment by ungeahnt** [comment]
Hi @zulus 

> For now I'm alone. 

Is there a reason for this? 

> I'm trying to split things and publish :/

Do you want to split the project so that others can take over and develop certain parts, or do you mean that you will develop and publish parts?

**Comment by zulus** [comment]
OK sorry guys I wrote it bad:

*I'm alone*: 
I mean I'm now only active developer i think

*I'm trying to split things and publish*:
Originally I wanted to drop WTP and DLTK and switch WWD and Handly. Because most of my time I spent on creating bypass between two of previous and/or creating workaround for they limitations

Unfortunately due my personal problems (I was on the verge of bankryptcy) I had to abandon my experiments and focus on daily (and night) work. Later because I need PHP8 I started work on basic PHP8 support on current code base. But due my ADHD now I have incredibly mess on disk (repository). 

I'm working on: take what works and put to master. My time is very limited, as I see I still need couple days to put on master branch.

Later I'll prepare "next" branch for WWD + Handly version and ask for some help, mostly for older PHP versions (maybe is time to reduce PDT to PHP >=7 ?)

Sorry for your nerves :/

**Comment by stanolacko** [comment]
Yes. It's solution to drop PHP <5.x support and focus on PHP 7>= or that
PHP 7.4 is no longer supported to only support PHP 8 >= and keep current
PDT as is.
If someone will need lower PHP support, he can create a separate workspace
with an older version of PDT.

**Comment by zulus** [comment]
About version reduction, why is important:
For now we have 3 parsers for each PHP versions:

1. Document structure (WTP integration)
1. DOM AST (editor colors, quick fix etc..)
1. DLTK AST (validation, indexing, type inference ...)

On next version I originally planned one parser for all PHP versions (latest) + extra validation directly on AST (similar to how PHP validate static expressions on constants), this will greatly reduce further maintenance, but for now wasn't finished.

Because at least DLTK AST have to be re-implemented, less versions === less work

**Comment by stanolacko** [comment]
PHP 7.x is well supported in previous versions. So I vote just for support
8.x.
https://www.php.net/supported-versions.php
7.4 reach the end of life.
So why support it in the new version?

**Comment by DarkNami** [comment]
Because Eclipse PDT is a IDE, not a compiler or VM ?

If developer can run 7.x source code (maximum used 7.x syntax/functions) on PHP 8.1 environment, but he/she can't maintenace source code by some PHP IDE. 

Or, some PHP IDE will break if a project included some 7.x packages from composer.

Is not it sound strange ?

**Comment by stanolacko** [comment]
Maybe.
But there is PHP 8 in production for 2 years for testers more.
8.0 is now in security fixes only and reaching EOL.
8.2 is now current PHP version.

And backward compatibility is something which is a "brake" to implement PHP
8.x into PDT.

So my practice is. If some library is not working in requested PHP version,
I will make own fork and fix it.
Maybe this will speed up that update process, which I remember from 5.6 to
7.x :)

**Comment by DarkNami** [comment]
Maybe Language Server Protocol is a optional solution ? 
https://langserver.org/

(not sure, I don't familiar with PDT source code.)

**Comment by innologi** [comment]
Hi,

> If developer can run 7.x source code (maximum used 7.x syntax/functions) on PHP 8.1 environment, but he/she can't maintenace source code by some PHP IDE.

If source code can run without issues in both 7.x and 8.1 environments, how could it not be compatible to a PHP 8.1 parser in PDT?

> Or, some PHP IDE will break if a project included some 7.x packages from composer.

I don't understand. You would not set up your project's vendor packages to be interpreted/parsed as your project source code, so why would they break a PHP 8+ parser in PDT?


I would add my vote for dropping parsers for officially unsupported PHP versions. It's still possible to revert to older PDT versions if you need to maintain code that is not compatible with PHP 8+.

**Comment by DarkNami** [comment]
I may have misunderstood some things.
I thought parser was incrementally, but parser are each PHP versions independent in fact ?

(I thought single parser support base syntax (7.0) + 7.1 syntax (incremental) + 7.2 syntax (incremental) + ...)

**Comment by spider4216** [comment]
Happy holidays everyone :christmas_tree::christmas_tree::christmas_tree: Hope we'll see PDT in 2023 with PHP 8 on board :pray::pray::pray:
It's my letter :envelope: to Santa Claus :santa:

**Comment by zulus** [comment]
I will need some help in upcoming months in testing if anyone  is still interested

**Comment by NainzguL** [comment]
sure don't hesitate to reach me if needed. Not sure how much time I can allow you but I'll do my best !

**Comment by wakabadou** [comment]
what should we do?

**Comment by pounard** [comment]
I am interested as well, I'm not afraid using Eclipse in a broken state (git will always save my life anyway) !

**Comment by zulus** [comment]
For now I upgraded my machine, and updated everything to make sure PDT is working correctly with recent Eclipse (2022-12) and all tests passes.

Next step is to finalize my experiments (from time before my ADHD + own company freezes me completely) and push to repository one by one. 

After that I'll ask You to test ;)

**Comment by merks** [comment]
If there's anything releng-related for which you need help, please reach out to be personally.

**Comment by stanolacko** [comment]
Cool! Thanks @zulus !

**Comment by spider4216** [comment]
Looking forward release :rocket:

**Comment by aln-1** [comment]
> I will need some help in upcoming months in testing if anyone is still interested

Count me in, I'll be happy to test.

**Comment by fbastien** [comment]
Thanks! I would be glad to help testing as well :)

**Comment by zulus** [comment]
I recently uploaded nearly final version of PHP 8.0 compiler parser, so basic syntax validation should be fine now. Semantic highlighting and code formatting not working correctly

Supported 8.0 syntax:
 
1. UnionTypes
2. Match expression
3. Named parameters
4. Attributes
5. Nullsafe operator
6. Static return type
7. Constructor property promotion

Some smaller PHP8.0 syntax changes have to be still implemented/verified on compiler parser first before I move code to DOM parser:

1. Throw expressions
2. Non capturing catches
3. Variable syntax consistency
4. Namespaced names consistency

**Comment by stanolacko** [comment]
Bravo.

Dňa po 17. 4. 2023, 21:24 Dawid Pakuła ***@***.***>
napísal(a):

**Comment by pounard** [comment]
@zulus Can we get it using something like a nightly update site or such ?

**Comment by zulus** [comment]
> @zulus Can we get it using something like a nightly update site or such ?

Yeah, builds are done via: https://ci.eclipse.org/pdt/view/PDT/job/pdt-master/

And available under: https://download.eclipse.org/tools/pdt/updates/latest-nightly/

Recent updates in compiler parser:

1. Throw expressions
2. Non capturing catches
3. Variable syntax consistency

One last thing on compiler parser level: Namespace names consistency. This is will by tricky, to keep backward compatibility with old tests

**Comment by aln-1** [comment]
> > @zulus Can we get it using something like a nightly update site or such ?
> 
> Yeah, builds are done via: https://ci.eclipse.org/pdt/view/PDT/job/pdt-master/
> 
> And available under: https://download.eclipse.org/tools/pdt/updates/latest-nightly/
> 
> Recent updates in compiler parser:
> 
> 1. Throw expressions
> 2. Non capturing catches
> 3. Variable syntax consistency
> 
> One last thing on compiler parser level: Namespace names consistency. This is will by tricky, to keep backward compatibility with old tests

Hi,
I just installed the nightly build. Here is some quick feedback.
The following syntaxes give fault errors:
* enum
* readonly class
* constructor variable with visibility (private, protected ..). Also use statements for variable classes are marked as not needed (yellow).
* use of trait in class gives "cannot be resolved"
* use "as" can't be resolved

**Comment by zulus** [comment]
For now im working on parsers, first for 8.0 (enum is from 8.1) ;) : https://www.php.net/releases/8.0/en.php

Validation will be, once I finish all parsers (8.0, 8.1, 8.2)

**Comment by stanolacko** [comment]
@zulus Excellent job! Made some few tests and 8.0 works nice.
What is missing for me, is autocomplete named attributes of functions, which can be on top of options.
Do you plan support of this help for developers?
For example setCookie example on PHP 8.0.
`setCookie(name: 'Example',` `CTRL-Space` and show all agruments to find for example `expires`

**Comment by spider4216** [comment]
> > @zulus Can we get it using something like a nightly update site or such ?
> 
> Yeah, builds are done via: https://ci.eclipse.org/pdt/view/PDT/job/pdt-master/
> 
> And available under: https://download.eclipse.org/tools/pdt/updates/latest-nightly/
> 
> Recent updates in compiler parser:
> 
>     1. Throw expressions
> 
>     2. Non capturing catches
> 
>     3. Variable syntax consistency
> 
> 
> One last thing on compiler parser level: Namespace names consistency. This is will by tricky, to keep backward compatibility with old tests

Thank's @zulus 
Install it and start to testing

**Comment by vardumper** [comment]
<img width="488" alt="Screenshot 2023-04-19 at 13 09 44" src="https://user-images.githubusercontent.com/21208397/233057249-c014af07-90fc-4e3b-b340-00eb8dd9c5c0.png">

I also test with lated with latest nightly and [experimental] php8.2 and saw that mixed is not recognised as a variable type.

**Comment by zulus** [comment]
I made basic infa for all recent PHP versions: https://github.com/eclipse/pdt/commit/fef3299de75bf535e2d08afffa50f628915c0b2b

Once I finish DOM parser for PHP8.0 I'll incrementally continue on 8.1

**Comment by zulus** [comment]
I made umbrella tasks for each PHP version for easier tracking

* #148
* #149 
* #150 

I've done PHP8.0 compiler parser : https://github.com/eclipse/pdt/commit/3de174d95aeb6c097f0aa2bc11f29253b4a26547

Once I finish DOM AST nodes (they are much more complicated than compiler by design), I'll prepare DOM AST parser

I'm closing this task for now

**Comment by spider4216** [comment]
I'm using generate getters and setters functional, but it was generated without any type (arguments and return values). Can you fix it somehow? Thanks

![image](https://user-images.githubusercontent.com/10693428/233542953-447bba0c-f94b-4076-b0ca-70d0ed7651ed.png)
